WordPress/Classes/ViewRelated/Domains/Domain registration/RegisterDomainSuggestions/RegisterDomainCoordinator.swift

```diff
@@ -0,0 +1,195 @@
+            title: title)
+        let navController = LightNavigationController(rootViewController: webViewController)
+
+        // WORKAROUND: The reason why we have to use this mechanism to detect success and failure conditions
+        // for domain registration is because our checkout process (for some unknown reason) doesn't trigger
+        // call to WKWebViewDelegate methods.
+        //
+        // This was last checked by @diegoreymendez on 2021-09-22.
+        //
+        webViewURLChangeObservation = webViewController.webView.observe(\.url, options: .new) { [weak self] _, change in
+            guard let self = self,
+                  let newURL = change.newValue as? URL else {
+                return
+            }
+
+            self.handleWebViewURLChange(newURL, domain: domainSuggestion.domainName, onCancel: {
+                if shouldPush {
+                    viewController.navigationController?.popViewController(animated: true)
+                } else {
+                    navController.dismiss(animated: true)
+                }
+            }) { domain in
+                viewController.dismiss(animated: true, completion: { [weak self] in
+                    guard let self = self else {
+                        return
+                    }
+
+                    self.domainPurchasedCallback?(viewController, domain)
+                })
+            }
+        }
+
+        if let site {
+            WPAnalytics.track(.domainsPurchaseWebviewViewed, properties: WPAnalytics.domainsProperties(for: site), blog: site)
+        } else {
+            // TODO: Track showing no site checkout
+        }
+
+        webViewController.configureSandboxStore {
+            if shouldPush {
+                viewController.navigationController?.pushViewController(webViewController, animated: true)
+            } else {
+                viewController.present(navController, animated: true)
+            }
+        }
+    }
+
+    /// Handles URL changes in the web view.  We only allow the user to stay within certain URLs.  Falling outside these URLs
+    /// results in the web view being dismissed.  This method also handles the success condition for a successful domain registration
+    /// through said web view.
+    ///
+    /// - Parameters:
+    ///     - newURL: the newly set URL for the web view.
+    ///     - domain: the domain the user is purchasing.
+    ///     - onCancel: the closure that will be executed if we detect the conditions for cancelling the registration were met.
+    ///     - onSuccess: the closure that will be executed if we detect a successful domain registration.
+    ///
+    private func handleWebViewURLChange(
+        _ newURL: URL,
+        domain: String,
+        onCancel: () -> Void,
+        onSuccess: (String) -> Void) {
+
+        let canOpenNewURL = newURL.absoluteString.starts(with: Constants.checkoutWebAddress)
+
+        guard canOpenNewURL else {
+            onCancel()
```

Comment:
[Suggestion] There is a critical bug in this implementation. Many times when I was testing, sometimes the default account authentication fails and the Log In page shows up instead of the Checkout page. And since the Log In url doesn't start with `Constants.checkoutWebAddress`, the web view is dismissed immediately.

Reply:
Interesting! Any idea how I can reproduce this?

Reply:
@hassaanelgarem I couldn't figure out a way to consistently reproduce that issue, but I remember it happened to many times during development ( using proxy ) and a fewer times in production ( without proxy ).

What do you say we start by logging this event on Sentry or Tracks instead of silently dismissing the Checkout screen?
 
We can iterate on this later if we see that this happens very frequently.

Also, when logging this event, we should also include the URL that caused the web view to be dismissed.

Reply:
As discussed on Slack, we suspect this to be a proxy issue. I am hesitant to add _error_ tracking for this, as this is not technically an error. It's only an error if the user is redirected to log in, but in other cases, it is an expected behavior. What do you think?